modules/util/source-formatter/src/main/java/com/liferay/source/formatter/util/PortalBootstrapBNDUtil.java

```diff
@@ -0,0 +1,178 @@
+import java.util.jar.JarEntry;
+import java.util.jar.JarInputStream;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * @author Hugo Huijser
+ */
+public class PortalBootstrapBNDUtil {
+
+	public static String updateExportPackages(File portalDir, String content)
+		throws IOException {
+
+		Map<String, String> packageVersionsMap = _getPackageVersionsMap(
+			portalDir, "lib/portal/", "lib/development/");
+
+		Matcher matcher = _exportVersionPattern.matcher(content);
+
+		while (matcher.find()) {
+			String packageName = matcher.group(1);
+
+			String version = packageVersionsMap.get(packageName);
+
+			if ((version == null) && packageName.endsWith(".*")) {
+				packageName = packageName.substring(
+					0, packageName.length() - 2);
+
+				for (Map.Entry<String, String> entry :
+						packageVersionsMap.entrySet()) {
+
+					String key = entry.getKey();
+
+					if (key.startsWith(packageName)) {
+						version = entry.getValue();
+
+						break;
+					}
+				}
+			}
+
+			if (version == null) {
+				continue;
+			}
+
+			String match = matcher.group();
+
+			String replacement = StringUtil.replaceLast(
+				match, matcher.group(2), version);
+
+			content = StringUtil.replace(content, match, replacement);
+		}
+
+		return content;
+	}
+
+	private static Map<String, String> _getPackageVersionsMap(
+			File portalDir, String... dirNames)
+		throws IOException {
+
+		Map<String, String> packageVerionsMap = new LinkedHashMap<>();
+
+		for (String dirName : dirNames) {
+			File dependenciesPropertiesFile = new File(
+				portalDir, dirName + "dependencies.properties");
+
+			if (!dependenciesPropertiesFile.exists()) {
+				continue;
+			}
+
+			Properties properties = new Properties();
+
+			properties.load(new FileInputStream(dependenciesPropertiesFile));
+
+			for (Map.Entry<Object, Object> entry : properties.entrySet()) {
+				String version = _getVersion((String)entry.getValue());
+
+				if (version == null) {
+					continue;
+				}
+
+				String key = (String)entry.getKey();
+
+				File jarFile = new File(portalDir, dirName + key + ".jar");
+
+				if (!jarFile.exists()) {
+					continue;
+				}
+
+				JarInputStream jarInputStream = null;
+
+				try {
+					jarInputStream = new JarInputStream(
+						new FileInputStream(jarFile));
+
+					while (true) {
+						JarEntry jarEntry = jarInputStream.getNextJarEntry();
+
+						if (jarEntry == null) {
+							break;
+						}
+
+						String name = jarEntry.getName();
+
+						if (name.endsWith(StringPool.SLASH) ||
+							!name.contains(StringPool.SLASH)) {
+
+							continue;
+						}
+
+						String packageName = StringUtil.replace(
+							name.substring(0, name.lastIndexOf(CharPool.SLASH)),
+							CharPool.SLASH, CharPool.PERIOD);
+
+						if (!packageVerionsMap.containsKey(packageName)) {
+							packageVerionsMap.put(packageName, version);
+						}
+					}
+				}
+				finally {
+					if (jarInputStream != null) {
+						jarInputStream.close();
+					}
+				}
```

Comment:
This logic should be a fallback if the jar is not an OSGi bundle already.

Several of our dependencies are OSGi bundles and the package versions need to be respected. So 

- first you should load the using `aQute.bnd.osgi.Domain.domain(File)`
- if `aQute.bnd.osgi.Domain.getBundleSymbolicName()` does not return `null` you know it's an OSGi bundle, otherwise use existing logic
- otherwise use the method `aQute.bnd.osgi.Domain.getExportPackage()` which returns a `aQute.bnd.header.Parameters` which has a method `aQute.bnd.header.Parameters.entrySet()` to iterate over the `keys` which are the package names, and value is a map containing a field `version` holding the package version

Reply:
THanks for the feedback, will look into that this week